internal/handshake/crypto_stream_conn.go

```diff
@@ -0,0 +1,99 @@
+package handshake
+
+import (
+	"bytes"
+	"io"
+	"net"
+	"time"
+)
+
+// The CryptoStreamConn is used as the net.Conn passed to mint.
+// It has two operating modes:
+// 1. It can read and write to bytes.Buffers.
+// 2. It can use a quic.Stream for reading and writing.
+// The buffer-mode is only used by the server, in order to statelessly handle retries.
+type CryptoStreamConn struct {
+	remoteAddr net.Addr
+
+	// the buffers are used before the session is initialized
+	readBuf  bytes.Buffer
+	writeBuf bytes.Buffer
+
+	// stream will be set once the session is initialized
+	stream io.ReadWriter
+}
+
+var _ net.Conn = &CryptoStreamConn{}
+
+// NewCryptoStreamConn creates a new CryptoStreamConn
+func NewCryptoStreamConn(remoteAddr net.Addr) *CryptoStreamConn {
+	return &CryptoStreamConn{remoteAddr: remoteAddr}
+}
+
+func (c *CryptoStreamConn) Read(b []byte) (int, error) {
+	if c.stream != nil {
+		return c.stream.Read(b)
+	}
+	return c.readBuf.Read(b)
+}
+
+// AddDataForReading adds data to the read buffer.
+// This data will ONLY be read when the stream has not been set.
+func (c *CryptoStreamConn) AddDataForReading(data []byte) {
+	c.readBuf.Write(data)
+}
+
+func (c *CryptoStreamConn) Write(p []byte) (int, error) {
+	if c.stream != nil {
+		return c.stream.Write(p)
+	}
+	return c.writeBuf.Write(p)
+}
+
+// GetDataForWriting returns all data currently in the write buffer, and resets this buffer.
+func (c *CryptoStreamConn) GetDataForWriting() []byte {
+	defer c.writeBuf.Reset()
+	return c.writeBuf.Bytes()
```

Comment:
You have to copy the slice here I think:

> The slice is valid for use only until the next buffer modification (that is, only until the next call to a method like Read, Write, Reset, or Truncate). 

https://golang.org/pkg/bytes/#Buffer.Bytes

Reply:
Good point, I missed that.